### DISCUSSION TITLE
Policies for Storage

### DISCUSSION THREAD

**Discussion by rbkayz** [supabase/supabase#7269] [Questions] [answered]
Team,

Thank you for being super helpful with all my banal questions :)

I'd like to understand more on how to enable specific access to files in storage for users based on policies. 

So, for example, here's a simple file I've uploaded to a bucket. Then I clicked on 'copy URL' to receive this - https://welladjldntnorgqjmvm.supabase.co/storage/v1/object/sign/message-body/sample.html?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1cmwiOiJtZXNzYWdlLWJvZHkvc2FtcGxlLmh0bWwiLCJpYXQiOjE2NTUxOTkzMzcsImV4cCI6MTk3MDU1OTMzN30.zxiDX0P-J-7UWw3Pe_DLMt-x8gnkLd1rOQq8TGizY6Y

What is this token? Checking on JWT.io shows this as an invalid signature

At the same time, if I replace this with my service_role token, it throws back a 400 access denied error

Is there some documentation where I can understand this better please

Thank you

**Comment by GaryAustin1** [comment, marked as answer]
The token is used for a signed URL and when it expires the url is no longer useable.  It is used for non-public buckets.

**Reply by rbkayz** [reply]
Oh, this is great. Thank you Gary. So if I were to only allow access to files with some conditions, say only a select set of users (based on conditions from another table) should be able to access a file, how can I do that?

And do I need to make the bucket public first?

**Reply by GaryAustin1** [reply]
You can restrict who can generate a signed URL with a security policy 
https://supabase.com/docs/reference/javascript/storage-from-createsignedurl
The security policy can use other tables.  

Once you have a SignedURL or a public URL (for a public bucket) you can not control who uses them.  They go straight to the S3 bucket and bypass security rules.

**Reply by shahbhavik01** [reply]
Hi @GaryAustin1  I have my supabase URL and supbase key. Is there documentation somewhere on what the policy to allow my supabase client to generate the signed URL should look like?

**Comment by rbkayz** [comment]
@GaryAustin1 this was great help. Thanks a ton. 

QQ - is there some tutorial on how to use a REST api to manage storage objects? I tried the swagger one, but it's throwing errors

Thank you

**Reply by GaryAustin1** [reply]
Besides resources listed here https://supabase.com/docs/guides/storage#resources and random examples on the web, I don't know of any.
You probably should close this one out and start a new question if you have a specific error or issue with Storage REST API.
